readme.md
# version

v0.0.1 - Baby steps

# Usage

1. api.sanskritworld.in/v0.0.1/verbform/verb
2. api.sanskritworld.in/v0.0.1/verbform/meaning
3. api.sanskritworld.in/v0.0.1/verbform/gana
4. api.sanskritworld.in/v0.0.1/verbform/number
5. api.sanskritworld.in/v0.0.1/verbform/madhaviya
6. api.sanskritworld.in/v0.0.1/verbform/kshiratarangini
7. api.sanskritworld.in/v0.0.1/verbform/dhatupradipa
8. api.sanskritworld.in/v0.0.1/verbform/uohyd
9. api.sanskritworld.in/v0.0.1/verbform/jnu
10. api.sanskritworld.in/v0.0.1/verbform/verbslp
11. api.sanskritworld.in/v0.0.1/verbform/lakara
12. api.sanskritworld.in/v0.0.1/verbform/upasarga
13. api.sanskritworld.in/v0.0.1/verbform/padadecider_id
14. api.sanskritworld.in/v0.0.1/verbform/padadecider_sutra
15. api.sanskritworld.in/v0.0.1/verbform/it_sutra
16. api.sanskritworld.in/v0.0.1/verbform/it_id
17. api.sanskritworld.in/v0.0.1/verbform/it_status
18. api.sanskritworld.in/v0.0.1/verbform/derivation
19. api.sanskritworld.in/v0.0.1/verbform/prakriya
20. api.sanskritworld.in/v0.0.1/verbform/prakriya/machine
21. api.sanskritworld.in/v0.0.1/verbform




changelog.md
# version

v0.0.1 - Baby steps
Date - 15 October 2017

## Usage

1. api.sanskritworld.in/v0.0.1/verbform/verb
2. api.sanskritworld.in/v0.0.1/verbform/meaning
3. api.sanskritworld.in/v0.0.1/verbform/gana
4. api.sanskritworld.in/v0.0.1/verbform/number
5. api.sanskritworld.in/v0.0.1/verbform/madhaviya
6. api.sanskritworld.in/v0.0.1/verbform/kshiratarangini
7. api.sanskritworld.in/v0.0.1/verbform/dhatupradipa
8. api.sanskritworld.in/v0.0.1/verbform/uohyd
9. api.sanskritworld.in/v0.0.1/verbform/jnu
10. api.sanskritworld.in/v0.0.1/verbform/verbslp
11. api.sanskritworld.in/v0.0.1/verbform/lakara
12. api.sanskritworld.in/v0.0.1/verbform/upasarga
13. api.sanskritworld.in/v0.0.1/verbform/padadecider_id
14. api.sanskritworld.in/v0.0.1/verbform/padadecider_sutra
15. api.sanskritworld.in/v0.0.1/verbform/it_sutra
16. api.sanskritworld.in/v0.0.1/verbform/it_id
17. api.sanskritworld.in/v0.0.1/verbform/it_status
18. api.sanskritworld.in/v0.0.1/verbform/derivation
19. api.sanskritworld.in/v0.0.1/verbform/prakriya
20. api.sanskritworld.in/v0.0.1/verbform/prakriya/machine
21. api.sanskritworld.in/v0.0.1/verbform


# v0.0.2

1. Generated files on ubuntu system. Windows is case insensitive.
Therefore earlier acIkarat and acIkaRat went into the same file.
See https://github.com/drdhaval2785/prakriya/issues/1
2. Prevented overwriting of data. See https://github.com/drdhaval2785/prakriya/issues/2.
3. In case of 404 errors, user is prompted the correct URI structure. See https://github.com/drdhaval2785/prakriya/issues/3.
4. Hosted API on api.sanskritworld.in. See https://github.com/drdhaval2785/prakriya/issues/5.
5. Changed webservice from flask to flask-restplus. See https://github.com/drdhaval2785/prakriya/issues/6.
6. Documented API via swagger, builtin in flask-restplus. See https://github.com/drdhaval2785/prakriya/issues/6#issuecomment-352167563.
7. Documented implementation details in swagger i.e. docstrings in methods.
See https://github.com/drdhaval2785/prakriya/issues/7
8. Return form info along with prakriyA.
See https://github.com/drdhaval2785/prakriya/issues/10
9. Enable input and output transliterations in API. See https://github.com/drdhaval2785/prakriya/issues/9
10. Corrected a typo in ASIrliN forms. See https://github.com/drdhaval2785/prakriya/issues/11
11. Corrected error in candrabindu of SLP1. See https://github.com/drdhaval2785/prakriya/issues/12
12. Started giving sUtra text for derivation in all APIs having prakriya in them. See https://github.com/drdhaval2785/prakriya/issues/13
13. Kept the intermediate and final forms as comma separated values instead of a list. See https://github.com/drdhaval2785/prakriya/issues/14
14. Changed the URI structure. See https://github.com/drdhaval2785/prakriya/issues/6#issuecomment-352253590
15. Input and output encodings are now changeable. See https://github.com/drdhaval2785/prakriya/issues/9.
16. Started using indic_transliteration package for transliteration



dhatu.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from flask import Flask, jsonify
import json
import os.path


app = Flask(__name__)
app.url_map.strict_slashes = False


@app.route('/v0.0.1/<string:verbform>', methods=['GET'])
def get_full_data(verbform):
    """Return the whole data regarding given verb form.

    This function reads the pregenerated JSON files.
    The files are created by the script at SanskritVerb repository.
    See https://github.com/drdhaval2785/SanskritVerb/blob/master/jsongenerator.sh for details.
    """
    uris = giveuris(verbform)
    fileofinterest = filepath(verbform)
    if os.path.exists(fileofinterest):
        with open(fileofinterest, 'r') as fin:
            return jsonify({'uris': uris, 'result': json.load(fin)})
    else:
        return jsonify({'uris': uris, 'filesearched': fileofinterest, 'error': 'The verb form is not in our database. If you feel it deserves to be included, kindly notify us on https://github.com/drdhaval2785/sktderivation/issues.'})


@app.route('/v0.0.1/<string:verbform>/prakriya', methods=['GET'])
def get_prakriya(verbform):
    """Return the specific sought for information of a given verb form."""
    fileofinterest = filepath(verbform)
    uris = giveuris(verbform)
    if not os.path.exists('/var/www/html/sanskritworldflask/data/sutrainfo.json'):
        return jsonify({'error': 'file data/sutrainfo.json missing. You can obtain it from https://github.com/drdhaval2785/SanskritVerb/blob/master/Data/sutrainfo.json'})
    elif os.path.exists(fileofinterest):
        with open(fileofinterest, 'r') as fin:
            verbdata = json.load(fin)
            result = []
            with open('/var/www/html/sanskritworldflask/data/sutrainfo.json', 'r') as sutrafile:
                sutrainfo = json.load(sutrafile)
            data = verbdata
            for datum in data:
                subresult = []
                for member in datum['derivation']:
                    subresult.append(sutrainfo[member['sutra_num']] + ' (' + member['sutra_num'] + ') -> ' + ','.join(member['text']))
                result.append(subresult)
            return jsonify({'uris': uris, 'message': 'for machine friendly output, use api.sanskritworld.in/v0.0.1/verbform/prakriya/machine', 'result': result})
    else:
        return jsonify({'uris': uris, 'error': 'The verb form is not in our database. If you feel it deserves to be included, kindly notify us on https://github.com/drdhaval2785/sktderivation/issues.'})


@app.route('/v0.0.1/<string:verbform>/prakriya/machine', methods=['GET'])
def get_prakriya_jsonified(verbform):
    """Return the specific sought for information of a given verb form."""
    uris = giveuris(verbform)
    fileofinterest = filepath(verbform)
    if not os.path.exists('/var/www/html/sanskritworldflask/data/sutrainfo.json'):
        return jsonify({'error': 'file data/sutrainfo.json missing. You can obtain it from https://github.com/drdhaval2785/SanskritVerb/blob/master/Data/sutrainfo.json'})
    elif os.path.exists(fileofinterest):
        with open(fileofinterest, 'r') as fin:
            verbdata = json.load(fin)
            result = []
            with open('/var/www/html/sanskritworldflask/data/sutrainfo.json', 'r') as sutrafile:
                sutrainfo = json.load(sutrafile)
            data = verbdata
            for datum in data:
                subresult = []
                for member in datum['derivation']:
                    subresult.append((sutrainfo[member['sutra_num']], member['sutra_num'], ','.join(member['text'])))
                result.append(subresult)
            return jsonify({'uris': uris, 'message': 'for human readable output, use api.sanskritworld.in/v0.0.1/verbform/prakriya', 'result': result})
    else:
        return jsonify({'uris': uris, 'error': 'The verb form is not in our database. If you feel it deserves to be included, kindly notify us on https://github.com/drdhaval2785/sktderivation/issues.'})


@app.route('/v0.0.1/<string:verbform>/<string:argument>', methods=['GET'])
def get_specific_info(verbform, argument):
    """Return the specific sought for information of a given verb form."""
    uris = giveuris(verbform)
    fileofinterest = filepath(verbform)
    with open(fileofinterest, 'r') as fin:
        verbdata = json.load(fin)
        tmp = []
        for datum in verbdata:
            arg = maparguments(argument)
            tmp.append(datum[arg])
        result = []
        for member in tmp:
            if member not in result:
                result.append(member)
        return jsonify({'uris': uris, 'result': result})


@app.errorhandler(404)
def not_found(error):
    """Give URIs if app hits 404."""
    uris = giveuris()
    return jsonify({'uris': uris, 'error': 'Use any of the URIs given below.'})


def filepath(verbform):
    """Return absolute path of file on server."""
    result = '/var/www/html/sanskritworldflask/data/json/' + verbform + '.json'
    return result


def giveuris(verbform='<verbform>', webserver='api.sanskritworld.in', version='v0.0.1'):
    """Give the URIs list for RESTful service."""
    uris = {'prakriya_human_readable': webserver + '/' + version + '/' + verbform + '/prakriya',
            'prakriya_machine_readable': webserver + '/' + version + '/' + verbform + '/prakriya/machine',
            'verb_devanagari': webserver + '/' + version + '/' + verbform + '/verb',
            'verb_meaning': webserver + '/' + version + '/' + verbform + '/meaning',
            'verb_number': webserver + '/' + version + '/' + verbform + '/number',
            'verb_gana': webserver + '/' + version + '/' + verbform + '/gana',
            'madhaviyadhatuvritti_link': webserver + '/' + version + '/' + verbform + '/madhaviya',
            'kshiratarangini_link': webserver + '/' + version + '/' + verbform + '/kshiratarangini',
            'dhatupradipa_link': webserver + '/' + version + '/' + verbform + '/dhatupradipa',
            'UoHyd_link': webserver + '/' + version + '/' + verbform + '/uohyd',
            'JNU_link': webserver + '/' + version + '/' + verbform + '/jnu',
            'verb_slp': webserver + '/' + version + '/' + verbform + '/verbslp',
            'lakAra_or_tense': webserver + '/' + version + '/' + verbform + '/lakara',
            'upasarga': webserver + '/' + version + '/' + verbform + '/upasarga',
            'padadecider_id': webserver + '/' + version + '/' + verbform + '/padadecider_id',
            'padadecider_sutra': webserver + '/' + version + '/' + verbform + '/padadecider_sutra',
            'it_sutra': webserver + '/' + version + '/' + verbform + '/it_sutra',
            'it_id': webserver + '/' + version + '/' + verbform + '/it_id',
            'it_status': webserver + '/' + version + '/' + verbform + '/it_status',
            'all_details': webserver + '/' + version + '/' + verbform,
            }
    return uris


def maparguments(argument):
    """Map api friendly arguments to actual JSON keys."""
    mapapi = {'verb': 'verb',
              'meaning': 'meaning',
              'number': 'number',
              'gana': 'gana',
              'madhaviya': 'mAdhavIya',
              'kshiratarangini': 'kzIratarangiNI',
              'dhatupradipa': 'dhAtupradIpa',
              'uohyd': 'UoHyd',
              'jnu': 'jnu',
              'verbslp': 'input',
              'lakara': 'lakAra',
              'upasarga': 'upasarga',
              'padadecider_id': 'padadecider_id',
              'padadecider_sutra': 'padadecider_sutra',
              'it_id': 'it_id',
              'it_sutra': 'it_sutra',
              'it_status': 'it_status'
              }
    return mapapi[argument]


if __name__ == '__main__':
    app.run()


dhatu_restplus.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Create a web service for verb form derivation.

See api.sanskritworld.in for documentation.
Author - Dr. Dhaval Patel
Date - 16 December 2017
email - drdhaval2785@gmail.com
"""
from flask import Flask, jsonify
from flask_restplus import Api, Resource, reqparse
from flask_cors import CORS
import os
import json
from indic_transliteration import sanscript


app = Flask(__name__)
CORS(app)
apiversion = 'v0.0.2'
api = Api(app, version=apiversion, title=u'Prakriyāpradarśinī API', description='Describes step by step verb form generation according to Paninian grammar. Please report any issues <a href="https://github.com/drdhaval2785/prakriya/issues">here</a>. For more details, please see the github projects for <a href="https://github.com/drdhaval2785/SanskritVerb">SanskritVerb</a> and <a href="https://github.com/drdhaval2785/prakriya">prakriya</a>.')
storagedirectory = '/var/www/html/sanskritworldflask/'


@api.route('/' + apiversion + '/verbforms/<string:input_transliteration>/<string:verbform>')
@api.doc(params={'verbform': 'Verb form to be examined.', 'input_transliteration': 'devanagari/slp1/iast/hk/wx/itrans/kolkata/velthuis'})
class FullData(Resource):
    """Return the JSON data regarding verb information and derivation steps for given verb form."""

    get_parser = reqparse.RequestParser()
    get_parser.add_argument('output_transliteration', location='args', default='devanagari', help='devanagari/slp1/iast/hk/wx/itrans/kolkata/velthuis')

    @api.expect(get_parser, validate=True)
    def get(self, input_transliteration, verbform):
        """Return the verb information and derivation steps for given verb form.

        This function reads the pregenerated JSON files.
        If the JSON file exists, it is loaded back.
        If the file is absent, an error is thrown.
        The JSON files are created by the script at SanskritVerb repository.
        See https://github.com/drdhaval2785/SanskritVerb/blob/master/jsongenerator.sh for details.
        """
        output_transliteration = self.get_parser.parse_args()['output_transliteration']
        if input_transliteration not in ['devanagari', 'slp1', 'iast', 'hk', 'wx', 'itrans', 'kolkata', 'velthuis']:
            return {'error': 'input_transliteration can only take specified values. Check again.'}
        if output_transliteration not in ['devanagari', 'slp1', 'iast', 'hk', 'wx', 'itrans', 'kolkata', 'velthuis']:
            return {'error': 'output_transliteration can only take specified values. Check again.'}
        verbform = sanscript.transliterate(verbform, input_transliteration, 'slp1')
        uris = giveuris(verbform)
        fileofinterest = filepath(verbform)
        if not os.path.exists(storagedirectory + 'data/sutrainfo.json'):
            return {'error': 'file data/sutrainfo.json missing. You can obtain it from https://github.com/drdhaval2785/SanskritVerb/blob/master/Data/sutrainfo.json'}
        elif os.path.exists(fileofinterest):
            with open(fileofinterest, 'r') as fin:
                verbdata = json.load(fin)
                result = []
                with open(storagedirectory + 'data/sutrainfo.json', 'r') as sutrafile:
                    sutrainfo = json.load(sutrafile)
                data = verbdata
                for datum in data:
                    subresult = {}
                    derivationlist = []
                    for item in datum:
                        if item in ['gana', 'padadecider_id', 'padadecider_sutra', 'number', 'meaning', 'lakara', 'verb', 'it_status', 'it_sutra', 'purusha', 'vachana', 'upasarga', 'suffix']:
                            tmp = datum[item]
                            if item == 'verb':
                                tmp = tmp.replace('!', '~')
                            subresult[item] = sanscript.transliterate(tmp, 'slp1', output_transliteration)
                        elif item == 'derivation':
                            pass
                        else:
                            subresult[item] = datum[item]
                    for member in datum['derivation']:
                        sutratext = sanscript.transliterate(sutrainfo[member['sutra_num']], 'slp1', output_transliteration)
                        sutranum = sanscript.transliterate(member['sutra_num'].replace('~', '-'), 'slp1', output_transliteration)
                        form = sanscript.transliterate(member['form'].replace('@', 'u~'), 'slp1', output_transliteration)
                        derivationlist.append({'sutra': sutratext, 'sutra_num': sutranum, 'form': form})
                    subresult['derivation'] = derivationlist
                    result.append(subresult)
                return jsonify(result)
        else:
            return {'uris': uris, 'error': 'The verb form is not in our database. If you feel it deserves to be included, kindly notify us on https://github.com/drdhaval2785/prakriya/issues.'}


@api.route('/' + apiversion + '/verbforms/<string:input_transliteration>/<string:verbform>/prakriya')
@api.doc(params={'verbform': 'Verb form to be examined.', 'input_transliteration': 'devanagari/slp1/iast/hk/wx/itrans/kolkata/velthuis'})
class GetPrakriya(Resource):
    """Return human readable derivation of a given verb form."""

    get_parser = reqparse.RequestParser()
    get_parser.add_argument('output_transliteration', location='args', default='devanagari', help='devanagari/slp1/iast/hk/wx/itrans/kolkata/velthuis')

    @api.expect(get_parser, validate=True)
    def get(self, verbform, input_transliteration):
        """Return human readable derivation of a given verb form.

        This function converts the derivation data into human readable form.
        Typical line is 'Paninian rule (rule number) -> state of the verb form'.
        e.g. 'sArvaDAtukArDaDAtukayoH (7.3.84) -> Bo+a+tas'.

        Dependencies

        This function has a dependency on sutrainfo.json.
        sutrainfo.json has sutra number, sutra text key value pairs.
        The data stored in JSON does not have sutra text. it only has sutra number.
        sutra text is supplied from sutrainfo.json.
        See https://github.com/drdhaval2785/prakriya/blob/master/data/sutrainfo.json.
        """
        output_transliteration = self.get_parser.parse_args()['output_transliteration']
        if input_transliteration not in ['devanagari', 'slp1', 'iast', 'hk', 'wx', 'itrans', 'kolkata', 'velthuis']:
            return {'error': 'input_transliteration can only take specified values. Check again.'}
        if output_transliteration not in ['devanagari', 'slp1', 'iast', 'hk', 'wx', 'itrans', 'kolkata', 'velthuis']:
            return {'error': 'output_transliteration can only take specified values. Check again.'}
        verbform = sanscript.transliterate(verbform, input_transliteration, 'slp1')
        fileofinterest = filepath(verbform)
        uris = giveuris(verbform)
        if not os.path.exists(storagedirectory + 'data/sutrainfo.json'):
            return {'error': 'file data/sutrainfo.json missing. You can obtain it from https://github.com/drdhaval2785/SanskritVerb/blob/master/Data/sutrainfo.json'}
        elif os.path.exists(fileofinterest):
            with open(fileofinterest, 'r') as fin:
                verbdata = json.load(fin)
                result = []
                with open(storagedirectory + 'data/sutrainfo.json', 'r') as sutrafile:
                    sutrainfo = json.load(sutrafile)
                data = verbdata
                for datum in data:
                    subresult = []
                    for member in datum['derivation']:
                        sutratext = sanscript.transliterate(sutrainfo[member['sutra_num']], 'slp1', output_transliteration)
                        sutranum = sanscript.transliterate(member['sutra_num'].replace('~', '-'), 'slp1', output_transliteration)
                        form = sanscript.transliterate(member['text'].replace('@', 'u~'), 'slp1', output_transliteration)
                        subresult.append({'sutra': sutratext, 'sutra_num': sutranum, 'form': form})
                    result.append(subresult)
                return jsonify(result)
        else:
            return {'uris': uris, 'error': 'The verb form is not in our database. If you feel it deserves to be included, kindly notify us on https://github.com/drdhaval2785/prakriya/issues.'}


@api.route('/' + apiversion + '/verbforms/<string:input_transliteration>/<string:verbform>/<string:argument>')
@api.doc(params={'verbform': 'Verb form to be examined.', 'argument': 'See Implementation notes / docstring of GET method.', 'input_transliteration': 'devanagari/slp1/iast/hk/wx/itrans/kolkata/velthuis'})
class SpecificInfo(Resource):
    """Return the specific sought for information of a given verb form."""

    get_parser = reqparse.RequestParser()
    get_parser.add_argument('output_transliteration', location='args', default='devanagari', help='devanagari/slp1/iast/hk/wx/itrans/kolkata/velthuis')

    @api.expect(get_parser, validate=True)
    def get(self, verbform, argument, input_transliteration):
        """Return the specific sought for information of a given verb form.

        Valid arguments and expected output are as follows.

        "verb" - Return verb in Devanagari without accent marks.

        "verbaccent" - Return the verb with accent marks. This item is not changed irrespective of transliteration choices.

        "lakara" - Return the lakAra (tense / mood) in which this form is generated.

        "purusha" - Return the purusha in which this form is generated.

        "suffix" - Return the suffix by which this form is generated.

        "vachana" - Return the vachana in which this form is generated.

        "gana" - Return the gaNa (class) of the verb.

        "meaning" - Return meaning of the verb in SLP1 transliteration.

        "number" - Return number of the verb in dhAtupATha.

        "madhaviya" - Return link to mAdhaviyadhAtuvRtti. http://sanskrit.uohyd.ac.in/scl/dhaatupaatha is the home page.

        "kshiratarangini" - Return link to kSIrataraGgiNI. http://sanskrit.uohyd.ac.in/scl/dhaatupaatha is the home page.

        "dhatupradipa" - Return link to dhAtupradIpa. http://sanskrit.uohyd.ac.in/scl/dhaatupaatha is the home page.

        "jnu" - Return link to JNU site for this verb form. http://sanskrit.jnu.ac.in/tinanta/tinanta.jsp is the home page.

        "uohyd" - Return link to UoHyd site for this verb form. http://sanskrit.uohyd.ac.in/cgi-bin/scl/skt_gen/verb/verb_gen.cgi is the home page.

        "upasarga" - Return upasarga, if any. Currently we do not support verb forms with upasargas.

        "padadecider_id" - Return the rule number which decides whether the verb is parasmaipadI, AtmanepadI or ubhayapadI.

        "padadecider_sutra" - Return the rule text which decides whether the verb is parasmaipadI, AtmanepadI or ubhayapadI.

        "it_id" - Returns whether the verb is seT, aniT or veT, provided the form has iDAgama.

        "it_status" - Returns whether the verb form has iDAgama or not. seT, veT, aniT are the output.

        "it_sutra" - Returns rule number if iDAgama is caused by some special rule.
        """
        output_transliteration = self.get_parser.parse_args()['output_transliteration']
        verbform = sanscript.transliterate(verbform, input_transliteration, 'slp1')
        fileofinterest = filepath(verbform)
        validargs = ['jnu', 'uohyd', 'dhatupradipa', 'it_id', 'it_status', 'it_sutra', 'kshiratarangini', 'lakara', 'madhaviya', 'padadecider_id', 'padadecider_sutra', 'upasarga', 'verbaccent', 'gana', 'meaning', 'number', 'verb', 'purusha', 'vachana', 'upasarga', 'suffix']
        if argument not in validargs:
            return {'error': 'argument is invalid. It can take only jnu, uohyd, dhatupradipa, it_id, it_status, it_sutra, kshiratarangini, lakara, madhaviya, padadecider_id, padadecider_sutra, upasarga, verbaccent, gana, meaning, number, verb, purusha, vachana, upasarga, suffix values.'}
        with open(fileofinterest, 'r') as fin:
            verbdata = json.load(fin)
            tmp = []
            for datum in verbdata:
                item = datum[argument]
                if argument in ['gana', 'padadecider_id', 'padadecider_sutra', 'number', 'meaning', 'lakara', 'verb', 'it_status', 'it_sutra', 'purusha', 'vachana', 'upasarga', 'suffix']:
                    if argument == 'verb':
                        item = item.replace('!', '~')
                    tmp.append(sanscript.transliterate(item, 'slp1', output_transliteration))
                else:
                    tmp.append(item)
            result = []
            for member in tmp:
                if member not in result:
                    result.append(member)
            return jsonify(result)


@app.errorhandler(404)
def not_found(error):
    """Give URIs if app hits 404."""
    uris = giveuris()
    return jsonify({'uris': uris, 'error': 'Use any of the URIs given below.'})


def filepath(verbform):
    """Return absolute path of file on server."""
    global storagedirectory
    return storagedirectory + 'data/json/' + verbform + '.json'


def giveuris(verbform='<verbform>', webserver='https://api.sanskritworld.in', version=apiversion):
    """Give the URIs list for RESTful service."""
    uris = {'prakriya_human_readable': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '/prakriya?output_transliteration=<output_transliteration>',
            'prakriya_machine_readable': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '/prakriya/machine?output_transliteration=<output_transliteration>',
            'verb_devanagari': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '/verb?output_transliteration=<output_transliteration>',
            'verb_meaning': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '/meaning?output_transliteration=<output_transliteration>',
            'verb_number': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '/number?output_transliteration=<output_transliteration>',
            'verb_gana': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '/gana?output_transliteration=<output_transliteration>',
            'purusha': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '/purusha?output_transliteration=<output_transliteration>',
            'vachana': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '/vachana?output_transliteration=<output_transliteration>',
            'suffix': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '/suffix?output_transliteration=<output_transliteration>',
            'madhaviyadhatuvritti_link': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '/madhaviya?output_transliteration=<output_transliteration>',
            'kshiratarangini_link': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '/kshiratarangini?output_transliteration=<output_transliteration>',
            'dhatupradipa_link': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '/dhatupradipa?output_transliteration=<output_transliteration>',
            'UoHyd_link': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '/uohyd?output_transliteration=<output_transliteration>',
            'JNU_link': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '/jnu?output_transliteration=<output_transliteration>',
            'verb_slp': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '/verbslp?output_transliteration=<output_transliteration>',
            'lakAra_or_tense': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '/lakara?output_transliteration=<output_transliteration>',
            'upasarga': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '/upasarga?output_transliteration=<output_transliteration>',
            'padadecider_id': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '/padadecider_id?output_transliteration=<output_transliteration>',
            'padadecider_sutra': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '/padadecider_sutra?output_transliteration=<output_transliteration>',
            'it_sutra': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '/it_sutra?output_transliteration=<output_transliteration>',
            'it_id': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '/it_id?output_transliteration=<output_transliteration>',
            'it_status': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '/it_status?output_transliteration=<output_transliteration>',
            'all_details': webserver + '/' + version + '/verbforms/<input_transliteration>/' + verbform + '?output_transliteration=<output_transliteration>',
            }
    return uris


def maparguments(argument):
    """Map api friendly arguments to actual JSON keys."""
    mapapi = {'madhaviya': 'mAdhavIya',
              'kshiratarangini': 'kzIratarangiNI',
              'dhatupradipa': 'dhAtupradIpa',
              'uohyd': 'UoHyd',
              'verbslp': 'input',
              'lakara': 'lakAra',
              }
    if argument in mapapi:
        return mapapi[argument]
    else:
        return argument


def reversemaparguments(jsonkey):
    """Reversemap json keys to arguments of API."""
    reversemapapi = {'mAdhavIya': 'madhaviya',
                     'kzIratarangiNI': 'kshiratarangini',
                     'dhAtupradIpa': 'dhatupradipa',
                     'UoHyd': 'uohyd',
                     'input': 'verbslp',
                     'lakAra': 'lakara',
                     }
    if jsonkey in reversemapapi:
        return reversemapapi[jsonkey]
    else:
        return jsonkey


if __name__ == '__main__':
    # app.run(debug=True)
    app.run()
